jetbrains-core/src/software/aws/toolkits/jetbrains/utils/AsyncEvaluator.kt

```diff
@@ -0,0 +1,110 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.utils
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.progress.ProcessCanceledException
+import org.jetbrains.annotations.TestOnly
+import org.jetbrains.concurrency.AsyncPromise
+import org.jetbrains.concurrency.Promise
+import org.jetbrains.concurrency.isPending
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import software.aws.toolkits.core.utils.info
+import java.util.concurrent.TimeUnit
+
+abstract class AsyncEvaluator<TEntry, TReturn> {
```

Comment:
I suggest to clarify the name of this class: let's name it `CachingAsyncEvaluator`?